Add all remaining sections to ITT monthly report

tomas-stefano opened the pull request:
## Context

This PR adds all remaining sections in the interface, when generating the ITTMonthlyReport.

Also refactor the headline statistics to look into the locales because we need to find the right attribute for each status.

I chosen to map into the same locale as configuration.

## Guidance to review

1. Does it work?
2. Does it produces the right data?

## Documentation to review

I added a documentation on https://github.com/DFE-Digital/apply-for-teacher-training/blob/e17a73f04aab7762f80f86a2ac4ac01df07ced7f/docs/bigquery_and_reports.md

Please feel free to add more or suggest any edits.

## Link to Trello card

https://trello.com/c/BUyDUGYX/917-monthly-statistics-candidate-age-group-section

tomas-stefano commented on config/locales/publications/itt_monthly_report_generator.yml:
```diff
@@ -1,23 +1,51 @@
           title: Offer conditions not met
+          application_metrics_column:
+            this_cycle: number_of_candidates_who_did_not_meet_any_offer_conditions_this_cycle_to_date
+            last_cycle: number_of_candidates_who_did_not_meet_any_offer_conditions_this_cycle_to_same_date_previous_cycle
```
@avinhurry I had to change slightly the locale for the new section. I think I am reaching a more stable format now ☝️

tomas-stefano commented on app/lib/dfe/bigquery/application_metrics.rb:
```diff
@@ -41,6 +42,22 @@ def self.candidate_headline_statistics(cycle_week:)
+            AND subject_filter_category = "Total excluding Further Education"
+            AND nonsubject_filter_category = "Age group"
+            ORDER BY nonsubject_filter ASC
+          SQL
```
I am still using raw SQL here. Let me know how you would prefer to tackle the two SQLs that we have on this class.
I thought to add a small DSL like:

```ruby
results = ::DfE::Bigquery.client.query(
  where(recruitment_cycle_year:, cycle_week:, subject_filter_category: 'Total excluding Further Education',  nonsubject_filter_category: 'Age group').order(nonsubject_filter: :asc).to_sql
)
```

But then I will have to create a small Arel methods which I am not sure if will pay the benefits. Let me know your thoughts!

JR-G replied:
I would be interested in moving more towards this approach – The queries are almost identical and we will have six of these!

tomas-stefano replied:
done! I [didn't make work with ](https://ukgovernmentdfe.slack.com/archives/C05D5S8C6PJ/p1699542530358379)Arel::Table  so I had to create a small and simple approach

tomas-stefano commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Monthly statistics: Candidate age group section](https://trello.com/c/BUyDUGYX/917-monthly-statistics-candidate-age-group-section)

tomas-stefano commented on docs/bigquery_and_reports.md:
````diff
@@ -0,0 +1,39 @@
+
+```ruby
+  Publications::ITTMonthlyReportGenerator.new.explain
+```
````
I am terrible at documentation as you can see. Please give suggestions of better phrasing. 🙏

inulty-dfe replied:
You're not bad :smiley:

tomas-stefano commented on app/models/publications/itt_monthly_report_generator.rb:
```diff
@@ -16,12 +19,45 @@ def to_h
+        puts '=' * 40
+        # rubocop:enable Rails/Output
+      end; nil
+    end
```
I decided to add the explain method to maintain a mental sanity around each query we do on BigQuery and each section on the report. Firstly I had the explain method on `DfE::Bigquery::ApplicationMetrics` but then I moved to here because conceptually it makes sense to `show all the queries from this report` to be on the report itself.

tomas-stefano replied:
Done! Renamed to describe

inulty-dfe commented on docs/bigquery_and_reports.md:
```diff
@@ -0,0 +1,39 @@
+
+## ITT monthly report
+
+We generate the ITT monthly report by making queries to big query and
```
```suggestion
We generate the ITT monthly report by making queries to BigQuery and
```

tomas-stefano replied:
Done!

inulty-dfe commented on docs/bigquery_and_reports.md:
```diff
@@ -0,0 +1,39 @@
+## ITT monthly report
+
+We generate the ITT monthly report by making queries to big query and
+then generate a JSON compiling the big query responses into a single structure.
```
```suggestion
then generate a JSON compiling the BigQuery responses into a single structure.
```

tomas-stefano replied:
Done

inulty-dfe commented on docs/bigquery_and_reports.md:
````diff
@@ -0,0 +1,39 @@
+  Publications::ITTMonthlyReportGenerator.new.to_h
+```
+
+In case any day the report fails for generate (e.g BigQuery is timing out, etc)
````
```suggestion
In case any day the report fails to generate (e.g BigQuery is timing out, etc)
```

tomas-stefano replied:
Done

inulty-dfe commented on docs/bigquery_and_reports.md:
```diff
@@ -0,0 +1,39 @@
+
+In case any day the report fails for generate (e.g BigQuery is timing out, etc)
+you can generate the JSON again by passing a generation and publication date.
+The code will calculate the last sunday from the generation date and consider
```
```suggestion
The code will calculate the last Sunday from the generation date and consider
```

tomas-stefano replied:
Done

inulty-dfe commented on docs/bigquery_and_reports.md:
````diff
@@ -0,0 +1,39 @@
+  ).to_h
+```
+
+In order to know all the queries we make to big query for this report,
````
```suggestion
In order to know all the queries we make to BigQuery for this report,
```

tomas-stefano replied:
Done

inulty-dfe reviewed (changes requested):
Apart from  the inline coments, I feel `lib/dfe/bigquery/application_metrics.rb` is straddling layers of abstration.

I don't think the methods `def self.where` and `def self.age_group` should be defined in the same class. They are operating on different layers of abstraction.

Nice work building your own ORM :brain: :smile:

tomas-stefano commented on app/models/publications/itt_monthly_report_generator.rb:
```diff
@@ -31,51 +67,68 @@ def period
-          last_cycle: application_metrics.number_of_candidates_who_did_not_meet_any_offer_conditions_this_cycle_to_same_date_previous_cycle,
-        },
-      }
+      data.tap do
```
Done

tomas-stefano commented:
> Apart from the inline coments, I feel `lib/dfe/bigquery/application_metrics.rb` is straddling layers of abstration.
> 
> I don't think the methods `def self.where` and `def self.age_group` should be defined in the same class. They are operating on different layers of abstraction.
> 
> Nice work building your own ORM 🧠 😄

Thanks! Addressed all comments and move the where to another place.

inulty-dfe reviewed (approved):
:rocket:

vassyz commented on app/models/publications/itt_monthly_report_generator.rb:
```diff
@@ -16,12 +19,45 @@ def to_h
+        sex_query: sex_query(cycle_week:),
+        area_query: area_query(cycle_week:),
+        phase_query: phase_query(cycle_week:),
+      }.each do |key, value|
```
This already got merged, but you can skip the rubocop comments with:
```ruby
}.each do |key, value|
  Rails.logger.info "========= #{key.to_s.humanize} =========="
  Rails.logger.info value
  Rails.logger.info '=' * 40
end; nil
```
